### PR TITLE
fix a method cache matching issue with Type{T}

### PR DESCRIPTION
**Pull request by vtjnash**

### CODE REVIEW

File: test/core.jl [hunk 1]
```diff
@@ -90,6 +87,20 @@ h11840(::Type{T}) where {T<:Tuple} = '4'
...
+# when specificity is reversed
+j11840(::DataType) = '1'
+j11840(::Union{Type{T}, T}) where {T} = '2' # force cache to contain leaftypes
+@test j11840(Union{Tuple{Int32}, Tuple{Int64}}) == '2'
+@test j11840(Tuple{Union{Int32, Int64}}) == '1' # DataType more specific than Type
+
+# but show we can correctly match types with alternate equivalent representations
+k11840(::Type{Union{Tuple{Int32}, Tuple{Int64}}}) = '2'
+@test k11840(Tuple{Union{Int32, Int64}}) == '2'
+@test k11840(Tuple{Union{Int32, Int64}}) == '2'
```
Comment (JeffBezanson) [inline test/core.jl]: Looks like a duplicate line. Did you mean to write something else?

**Comment by JeffBezanson** [comment]
What do you mean by "keep track separately of what Kind it needs to match"? It sounds like that would refer to some extra piece of state, but I don't see one.

File: test/core.jl (hunk 1 above)
Comment (vtjnash) [reply test/core.jl]: Ned to repeat to use/check the cache

**Comment by vtjnash** [comment]
It already tracked the extra state, it just wasn’t setting it correctly in enough cases

File: test/core.jl (hunk 1 above)
Comment (StefanKarpinski) [reply test/core.jl]: Perhaps a comment to that effect then?

**Pull request merged**